AMDGPU/GlobalISel: Handle atomic sextload and zextload

arsenm opened the pull request:
Atomic loads are handled differently from the DAG, and have separate opcodes
and explicit control over the extensions, like ordinary loads. Add
new patterns for these.

There's room for cleanup and improvement. d16 cases aren't handled.

Fixes #111645

arsenm committed:
AMDGPU/GlobalISel: Handle atomic sextload and zextload

Atomic loads are handled differently from the DAG, and have separate opcodes
and explicit control over the extensions, like ordinary loads. Add
new patterns for these.

There's room for cleanup and improvement. d16 cases aren't handled.

Fixes #111645

arsenm commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/111721?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#111721** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/111721?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#111720** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/111720?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

rampitec commented on llvm/test/CodeGen/AMDGPU/GlobalISel/atomic_load_flat.ll:
```diff
@@ -0,0 +1,331 @@
+; GFX9-LABEL: atomic_load_flat_monotonic_i8_sext_to_i32:
+; GFX9:       ; %bb.0:
+; GFX9-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
+; GFX9-NEXT:    flat_load_sbyte v2, v[0:1] glc
```
What happened here? Why is there 2 loads?

arsenm replied:
There's a bug in amdgpu-postlegalizer-combiner, so it's not from the selection

rampitec reviewed (commented):
Missing test for buffer loads?

arsenm commented:
> Missing test for buffer loads?

Those are the gfx7 global cases. There aren't any atomic buffer load intrinsics

rampitec commented:
> > Missing test for buffer loads?
> 
> Those are the gfx7 global cases. There aren't any atomic buffer load intrinsics

But patch adds several MUBUF_Pseudo_Load_Pats which are not covered by tests?

arsenm commented:
> But patch adds several MUBUF_Pseudo_Load_Pats which are not covered by tests?

The only cases that might have missing coverage is extend to 16-bit register cases. In the DAG we didn't have legal 16-bit types on gfx6/7, but we could handle the loads here

arsenm commented:
ping

rampitec commented:
> > But patch adds several MUBUF_Pseudo_Load_Pats which are not covered by tests?
> 
> The only cases that might have missing coverage is extend to 16-bit register cases. In the DAG we didn't have legal 16-bit types on gfx6/7, but we could handle the loads here

I still think these added MUBUF patterns need at least some test coverage.

arsenm committed:
Remove i16 result mubuf types.

This would only be used if we started using direct i16
values on gfx6/gfx7, but they should legalize all to i32

arsenm commented:
> I still think these added MUBUF patterns need at least some test coverage.

Just removed the 16-bit result ones

rampitec commented on llvm/lib/Target/AMDGPU/BUFInstructions.td:
```diff
@@ -983,15 +983,20 @@ defm BUFFER_LOAD_LDS_U16 : MUBUF_Pseudo_Loads_LDSOpc <
 >;
 
 defm : MUBUF_Pseudo_Load_Pats<"BUFFER_LOAD_UBYTE", i32, atomic_load_8_global>;
+defm : MUBUF_Pseudo_Load_Pats<"BUFFER_LOAD_UBYTE", i32, atomic_load_zext_8_global>;
```
But these are still not covered?

arsenm committed:
Add missing gfx6 run line

arsenm committed:
Fix memory predicates

arsenm reviewed (commented):
ping

rampitec commented on llvm/lib/Target/AMDGPU/BUFInstructions.td:
```diff
@@ -983,15 +983,20 @@ defm BUFFER_LOAD_LDS_U16 : MUBUF_Pseudo_Loads_LDSOpc <
 >;
 
 defm : MUBUF_Pseudo_Load_Pats<"BUFFER_LOAD_UBYTE", i32, atomic_load_8_global>;
+defm : MUBUF_Pseudo_Load_Pats<"BUFFER_LOAD_UBYTE", i32, atomic_load_zext_8_global>;
```
I still think if you are adding buffer patterns you need to have tests for them.

arsenm replied:
They are there

rampitec replied:
I see modified tests for flat, global and local.

arsenm replied:
This is not for buffer intrinsics or fat pointers. This is for global pointers implemented with mubuf instructions. This is covered by the global run lines for gfx6. The -flat-for-global case for gfx7 do not correctly use the mubuf results, but that's a separate issue

rampitec replied:
Ok

arsenm commented:
### Merge activity

* **Oct 31, 10:38 AM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/111721).
